Modules/Core/Common/include/itkMetaDataObject.h

```diff
@@ -149,6 +149,24 @@ class ITK_TEMPLATE_EXPORT MetaDataObject : public MetaDataObjectBase
   ~MetaDataObject() override = default;
 
 private:
+  /** Tells whether the specified pointers point to objects that compare equal.
+   * \note The template parameter and the trailing return type are there, just to enable SFINAE.*/
+  template <typename TMetaDataObject>
+  static auto
+  EqualMetaDataObjects(const TMetaDataObject * const lhs, const TMetaDataObject * const rhs) -> decltype(*lhs == *rhs)
+  {
+    static_assert(std::is_same<TMetaDataObject, MetaDataObjectType>::value, "Check the template argument.");
+    return *lhs == *rhs;
+  }
+
+  /** Overload for MetaDataObject types that do not have an `operator==`.*/
```

Comment:
I fear that in the case that `operator==` does exist, this code will produce two versions of `EqualMetaDataObjects`.  If the testing code works on all platforms then those two versions are not considered redefinitions of each other, but it may be difficult to know which of the two the compiler is choosing.  If the testing is working then maybe my fears are unfounded.

If `EqualMetaDataObjects` were a class instead of a function then we could put the second version here as the default implementation and the first version here as a template specialization.

Reply:
> If the testing code works on all platforms then those two versions are not considered redefinitions of each other, but it may be difficult to know which of the two the compiler is choosing. If the testing is working then maybe my fears are unfounded.

Just to be 100% sure, I did specifically extend the existing `MetaDataDictionary.Equal` unit tests to also check comparing dictionaries after storing an `std::string`. Because `std::string` is an obvious example for which `operator==` is very different from a bitwise comparison (memcmp). Luckily, the tests are passing successfully 😃

https://github.com/InsightSoftwareConsortium/ITK/blob/d84e49daa45429f3b2a6acc7efc25bfdb3a1129e/Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx#L290